src/main/scala/chisel3/ModuleChoice.scala

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package chisel3
+
+import scala.language.experimental.macros
+import scala.collection.immutable.ListMap
+
+import chisel3.FixedIOBaseModule
+import chisel3.experimental.{BaseModule, SourceInfo}
+import chisel3.internal.{groupByIntoSeq, Builder, WireBinding}
+import chisel3.internal.Builder.pushCommand
+import chisel3.internal.firrtl.DefInstanceChoice
+import chisel3.internal.sourceinfo.InstChoiceTransform
+
+object ModuleChoice extends SourceInfoDoc {
+
+  /** A wrapper method for Module instantiation based on option choices.
+    * (necessary to help Chisel track internal state). @see [[OptionGroup]]
+    * on how to declare the options for the alternatives.
+    *
+    * Example:
+    * ```
+    * val module = ModuleChoice(new DefaultModule)(Seq(
```

Comment:
what is the scala type of `module`? Is this thing figuring out the scala-common type between them?

Reply:
It's going to be the type parameter of the `FixedIOModule`s used.